chain-core/src/property.rs

```diff
@@ -0,0 +1,228 @@
+//! * the block number, its position in the blockchain relative
+//!   to the beginning of the chain. We often call this number
+//!   the block Date.
+//!
+//! # HasTransaction and Transaction
+//!
+//! These traits are mainly fit for the purpose of the Unspent Transaction
+//! Output (UTxO) model.
+//!
+//! # Ledger
+//!
+//! this trait is to make sure we are following the Transactions of the chain
+//! appropriately.
+//!
+//! # LeaderSelection
+//!
+//! This trait is following the protocol of the blockchain is followed
+//! properly and determined a given instance of the LeaderSelection object
+//! is selected to write a block in the chain.
+//!
+
+/// Block property
+///
+/// a block is part of a chain of block called Blockchain.
+/// the chaining is done via one block pointing to another block,
+/// the parent block (the previous block).
+///
+/// This means that a blockchain is a link-list, ordered from the most
+/// recent block to the furthest/oldest block.
+///
+/// The Oldest block is called the Genesis Block.
+pub trait Block: Serializable {
+    /// the Block identifier. It must be unique. This mean that
+    /// 2 different blocks have 2 different identifiers.
+    ///
+    /// In bitcoin this block is a SHA2 256bits. For Cardano's
+    /// blockchain it is Blake2b 256bits.
+    type Id;
+
+    /// the block date (also known as a block number) represents the
+    /// absolute position of the block in the chain. This can be used
+    /// for random access (if the storage algorithm allows it) or for
+    /// identifying the position of a block in a given epoch or era.
+    type Date;
+
+    /// return the Block's identifier.
+    fn id(&self) -> Self::Id;
+
+    /// get the parent block identifier (the previous block in the
+    /// blockchain).
+    fn parent_id(&self) -> &Self::Id;
+
+    /// get the block date of the block
+    fn date(&self) -> &Self::Date;
+}
+
+/// define a transaction within the blockchain. This transaction can be used
+/// for the UTxO model. However it can also be used for any other elements that
+/// the blockchain has (a transaction type to add Stacking Pools and so on...).
+///
+pub trait Transaction: Serializable {
+    /// the input type of the transaction (if none use `()`).
+    type Input;
+    /// the output type of the transaction (if none use `()`).
+    type Output;
+    /// a unique identifier of the transaction. For 2 different transactions
+    /// we must have 2 different `Id` values.
+    type Id;
+
+    fn inputs<'a>(&'a self) -> std::slice::Iter<'a, Self::Input>;
+    fn outputs<'a>(&'a self) -> std::slice::Iter<'a, Self::Output>;
+
+    /// return the Transaction's identifier.
+    fn id(&self) -> Self::Id;
+}
+
+/// accessor to transactions within a block
+///
+/// This trait is generic enough to show there is multiple types
+/// of transaction possibles:
+///
+/// * UTxO
+/// * certificate registrations
+/// * ...
+pub trait HasTransaction<T: Transaction> {
```

Comment:
This name is a bit odd. But then, we never came around to sufficiently generalizing this API, so maybe we can redesign this later.

Reply:
you know what, isn't this type a bit redundant with `IntoIterator` ?

Reply:
I thought that too, but it's probably not very ergonomic unless `IntoIterator` is implemented for the block object's reference.

Reply:
or even, with `AsRef<[T]> where T: Transaction`

Reply:
There may be an elegant generic way there that also clicks together with other methods consuming an `IntoIterator<Item = &Transaction>`. I should look into that, but as this is mostly a code move, it does not have to be done here.